oauth2client/service_account.py

```diff
@@ -12,96 +12,265 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-"""A service account credentials class.
+"""oauth2client Service account credentials class."""
 
-This credentials class is implemented on top of rsa library.
-"""
 
 import base64
+import copy
 import datetime
 import json
 import time
 
-from pyasn1.codec.ber import decoder
-from pyasn1_modules.rfc5208 import PrivateKeyInfo
-import rsa
-
 from oauth2client import GOOGLE_REVOKE_URI
 from oauth2client import GOOGLE_TOKEN_URI
 from oauth2client._helpers import _json_encode
 from oauth2client._helpers import _from_bytes
-from oauth2client._helpers import _to_bytes
 from oauth2client._helpers import _urlsafe_b64encode
 from oauth2client import util
 from oauth2client.client import AssertionCredentials
 from oauth2client.client import EXPIRY_FORMAT
+from oauth2client.client import SERVICE_ACCOUNT
+from oauth2client import crypt
+
+
+_KEYFILE_MAPPING = (
+    ('client_email', 'service_account_email'),
+    ('private_key', 'private_key_pkcs8_pem'),
+    ('private_key_id', 'private_key_id'),
+    ('client_id', 'service_account_id'),
+)
+_PASSWORD_DEFAULT = 'notasecret'
+_AT_LEAST_ONE_KEY = ('Exactly one of private_key_pkcs8_pem and '
+                     'private_key_pkcs12 must be passed to the '
+                     'ServiceAccountCredentials constructor.')
+_PKCS12_ERROR = r"""
+This library only implements PKCS#12 support via the pyOpenSSL library.
+Either install pyOpenSSL, or please convert the .p12 file
+to .pem format:
+    $ cat key.p12 | \
+    >   openssl pkcs12 -nodes -nocerts -passin pass:notasecret | \
+    >   openssl rsa > key.pem
+"""
+_PKCS12_KEY = '_private_key_pkcs12'
+
+
+class ServiceAccountCredentials(AssertionCredentials):
+    """Service Account credential for OAuth 2.0 signed JWT grants.
+
+    Supports
+
+    * JSON keyfile (typically contains a PKCS8 key stored as
+      PEM text)
+    * ``.p12`` key (stores PKCS12 key and certificate)
+
+    Makes an assertion to server using a signed JWT assertion in exchange
+    for an access token.
+
+    This credential does not require a flow to instantiate because it
+    represents a two legged flow, and therefore has all of the required
+    information to generate and refresh its own access tokens.
+
+    At most one of ``private_key_pkcs8_pem`` and ``private_key_pkcs12``
+    can be set.
 
+    Args:
+        service_account_email: string, The email associated with the
+                               service account.
+        private_key_pkcs8_pem: bytes, The content of a PKCS#8 key in PEM
```

Comment:
That's what I thought - document it as "(Optional)" the same way the other optional parameters are documented?